control-plane/connect-inject/peering_dialer_controller.go

```diff
@@ -0,0 +1,281 @@
+		}
+	}
+
+	// If spec secret doesn't exist, error because we can only initiate peering if we have a token to initiate with.
+	if existingSpecSecret == nil {
+		err = errors.New("PeeringDialer spec.peer.secret does not exist")
+		_ = r.updateStatusError(ctx, peeringDialer, err)
+		return ctrl.Result{}, err
+	}
+
+	// Read the peering from Consul.
+	// TODO(peering): do we need to pass in partition?
+	readReq := api.PeeringReadRequest{Name: peeringDialer.Name}
+	r.Log.Info("reading peering from Consul", "name", peeringDialer.Name)
+	peering, _, err := r.ConsulClient.Peerings().Read(ctx, readReq, nil)
+	var statusErr api.StatusError
+	peeringExists := true
+	if errors.As(err, &statusErr) && statusErr.Code == http.StatusNotFound && peering == nil {
+		peeringExists = false
+	} else if err != nil {
+		r.Log.Error(err, "failed to get Peering from Consul", "name", req.Name)
+		return ctrl.Result{}, err
+	}
+	// TODO(peering): Verify that the existing peering in Consul is an dialer peer. If it is an acceptor peer, an error should be thrown.
+
+	// At this point, we know the spec secret exists. If the status secret doesn't
+	// exist, then we want to initiate peering and update the status with the secret for the token being used.
+	if existingStatusSecret == nil {
+		// Whether the peering exists in Consul or not we want to initiate the peering so the status can reflect the
+		// correct secret specified in the spec.
+		r.Log.Info("status.secret doesn't exist or wasn't set, establishing peering with the existing spec.peer.secret", "secret-name", peeringDialer.Spec.Peer.Secret.Name, "secret-namespace", peeringDialer.Namespace)
+		peeringToken := existingSpecSecret.Data[peeringDialer.Spec.Peer.Secret.Key]
+		_, err := r.initiatePeering(ctx, peeringDialer.Name, string(peeringToken))
+		if err != nil {
+			_ = r.updateStatusError(ctx, peeringDialer, err)
+			return ctrl.Result{}, err
+		} else {
+			err := r.updateStatus(ctx, peeringDialer, peeringToken)
+			return ctrl.Result{}, err
+		}
+	} else {
+		// At this point, the status secret does exist.
+		// If the peering in Consul does not exist, initiate peering.
+		if !peeringExists {
+			r.Log.Info("status.secret exists, but the peering doesn't exist in Consul; establishing peering with the existing spec.peer.secret", "secret-name", peeringDialer.Spec.Peer.Secret.Name, "secret-namespace", peeringDialer.Namespace)
+			peeringToken := existingSpecSecret.Data[peeringDialer.Spec.Peer.Secret.Key]
+			_, err := r.initiatePeering(ctx, peeringDialer.Name, string(peeringToken))
+			if err != nil {
+				_ = r.updateStatusError(ctx, peeringDialer, err)
+				return ctrl.Result{}, err
+			} else {
+				err := r.updateStatus(ctx, peeringDialer, peeringToken)
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Or, if the peering in Consul does exist, compare it to the contents of the spec's secret. If there's any
+		// differences, initiate peering.
+		if r.specStatusSecretsDifferent(peeringDialer, existingSpecSecret) {
```

Comment:
Can we compare the token in the spec secret against the peering that exists in Consul? Does the peering that we read from Consul on L121 have that token value? if it doesn't, this step makes sense.

Reply:
Ah no, the peering read from Consul doesn't have the token value in it-- it may have some elements, but the token contents could also change to include parts that are secret so we may not be able to get some fields.

Reply:
thanks for the explanation Nitya!